lib/dry/result_matcher/evaluator.rb

```diff
@@ -0,0 +1,40 @@
+module Dry
+  module ResultMatcher
+    class Evaluator
+      attr_reader :result
+      attr_reader :output
+
+      def initialize(result, cases)
+        @result = result
+        @output = nil
+        @matched = false
+
+        cases.each do |name, kase|
+          define_singleton_method(name) do |pattern = nil, &block|
+            handle_case(kase, pattern, &block)
+          end
+        end
+      end
+
+      def call(&block)
+        block.call(self)
+        output
+      end
+
+      private
+
+      def handle_case(kase, pattern, &block)
+        return(output) if matched?
+
+        if kase.matches?(pattern, result)
+          @matched = true
+          @output = block.call(kase.resolve(result))
+        end
```

Comment:
Yeah. I can't think of anything right now apart from bundling the match case methods along with some state to track the output and whether a not a match has already been made.

So are you happy for this to remain in place?